pyro/infer/abstract_infer.py

```diff
@@ -313,25 +313,23 @@ def _remove_dropped_nodes(self, trace):
                 continue
 
     def _adjust_to_data(self, trace, data_trace):
-        for name, site in list(trace.nodes.items()):
+        for name, site in trace.iter_stochastic_nodes():
             # Adjust subsample sites
             if site_is_subsample(site):
                 site["fn"] = data_trace.nodes[name]["fn"]
                 site["value"] = data_trace.nodes[name]["value"]
             # Adjust sites under conditionally independent stacks
-            try:
-                site["cond_indep_stack"] = data_trace.nodes[name]["cond_indep_stack"]
-                site["fn"] = data_trace.nodes[name]["fn"]
-                for cis in site["cond_indep_stack"]:
-                    # Select random sub-indices to replay values under conditionally independent stacks.
-                    # Otherwise, we assume there is an dependence of indexes between training data
-                    # and prediction data.
+            site["cond_indep_stack"] = data_trace.nodes[name]["cond_indep_stack"]
+            site["fn"] = data_trace.nodes[name]["fn"]
+            for cis in site["cond_indep_stack"]:
+                # Select random sub-indices to replay values under conditionally independent stacks.
+                # Otherwise, we assume there is an dependence of indexes between training data
+                # and prediction data.
+                if not site_is_subsample(site):
```

Comment:
Correct me if I'm mistaken, but I don't think you should ever get here if `site_is_subsample` is True.

Reply:
You are probably right, so I have changed the `if` to an assertion. 😄 
Maybe I should add a test with nested plates as well.